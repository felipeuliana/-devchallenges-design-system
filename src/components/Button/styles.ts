import styled, { css } from 'styled-components';
import { ButtonProps } from '.';

const _COLORS = new Map([
  ['danger', { regular: '#d32f2f', alternative: '#fff' }],
  ['default', { regular: '#e0e0e0', alternative: '#3f3f3f' }],
  ['primary', { regular: '#2962ff', alternative: '#fff' }],
  ['secondary', { regular: '#455a64', alternative: '#fff' }],
]);

const _SIZES = new Map([
  ['lg', '1.1rem 2.2rem'],
  ['md', '0.8rem 1.6rem'],
  ['sm', '0.6rem 1.2rem'],
]);

const _wrapperModifiers = {
  default: (color: string) => css`
    background-color: ${_COLORS.get(color)?.regular};
    border-color: ${_COLORS.get(color)?.regular};
    color: ${_COLORS.get(color)?.alternative};
  `,
  outline: () => css`
    background-color: ${_COLORS.get('primary')?.alternative};
    border-color: ${_COLORS.get('primary')?.regular};
    color: ${_COLORS.get('primary')?.regular};
    border-color: ${_COLORS.get(color)?.regular};
    color: ${_COLORS.get(color)?.regular};
  `,
  text: () => css`
    background-color: ${_COLORS.get('primary')?.alternative};
    border-color: ${_COLORS.get('primary')?.alternative};
    color: ${_COLORS.get('primary')?.regular};
    border-color: transparent;
    color: ${_COLORS.get(color)?.regular};
  `,
  setBoxShadow: () => css`
    box-shadow: 0 0.2rem 0.3rem rgba(51, 51, 51, 0.2);
  `,
  setDisabled: (hasBackground: boolean) => css`
    &[disabled] {
      background-color: ${hasBackground
        ? _COLORS.get('default')?.regular
        : _COLORS.get('primary')?.alternative};
      border-color: transparent;
      color: #9e9e9e;
      cursor: not-allowed;
    }
  `,
  setSize: (size: string) => css`
    padding: ${_SIZES.get(size)};
  `,
};

export const Wrapper = styled.button<ButtonProps>`
  ${({ color, disabled, disableShadow, size, variant }) => {
    const _enableShadow = variant === 'default' && !disableShadow && !disabled;

    return css`
      border: 0.1rem solid;
      border-radius: 0.6rem;
      cursor: pointer;
      font-family: 'Noto Sans JP', sans-serif;
      font-size: 1.4rem;
      font-style: normal;
      font-weight: 500;
      line-height: 20px;
      padding: 0.8rem 1.6rem;
      text-align: center;
      transition: background-color 0.25s ease-in-out;

      ${!!variant && !!color && _wrapperModifiers[variant](color)}
      ${!!size && _wrapperModifiers.setSize(size)}
      ${disabled && _wrapperModifiers.setDisabled(variant !== 'text')}
      ${_enableShadow && _wrapperModifiers.setBoxShadow()}
    `;
  }}
`;
