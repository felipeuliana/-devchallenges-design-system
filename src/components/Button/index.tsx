import React from 'react';
import * as S from './styles';

export type ButtonProps = {
  children: React.ReactNode;
  disableShadow?: boolean;
  variant?: 'outline' | 'text' | null;
};

const Button = ({ children, disableShadow, variant = null }: ButtonProps) => (
  <S.Wrapper disableShadow={disableShadow} variant={variant}>
    {children}
  </S.Wrapper>
);

export default Button;
