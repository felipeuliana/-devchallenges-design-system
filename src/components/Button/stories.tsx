import React from 'react';
import Button, { ButtonProps } from '.';
import { Story, Meta } from '@storybook/react/types-6-0';

export default {
  component: Button,
  title: 'Button',
} as Meta;

export const Default: Story<ButtonProps> = (args) => <Button {...args} />;

Default.args = {
  children: 'Button',
};
