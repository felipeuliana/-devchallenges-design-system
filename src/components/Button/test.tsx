import React from 'react';
import Button from '.';
import { render, screen } from '@testing-library/react';
import { InsertEmoticon } from '@styled-icons/material';

const DISABLED_STYLE = {
  'background-color': '#e0e0e0',
  'border-color': 'transparent',
  color: '#9e9e9e',
  'pointer-events': 'none',
};

describe('<Button />', () => {
  it('should render correctly', () => {
    const { container } = render(
      <>
        <Button
          color="primary"
          size="lg"
          leftIcon={<InsertEmoticon title="Emoticon" />}
        >
          Emoticon
        </Button>
        <Button as="a" href="/link" size="sm" variant="text">
          Link
        </Button>
      </>
    );

    expect(container).toMatchSnapshot();
  });

  it('should allow access to button attributes', () => {
    render(
      <Button disabled name="Default" value="Default">
        Default
      </Button>
    );

    const defaultButton = screen.getByRole('button', { name: /default/i });

    expect(defaultButton).toHaveAttribute('disabled');
    expect(defaultButton).toHaveAttribute('name');
    expect(defaultButton).toHaveAttribute('value');
  });

  describe('"default" type', () => {
    it('should be styled correctly', () => {
      render(<Button>Default</Button>);

      expect(screen.getByRole('button', { name: /default/i })).toHaveStyle({
        'background-color': '#e0e0e0',
        'box-shadow': '0 0.2rem 0.3rem rgba(51,51,51,0.2)',
        color: '#3f3f3f',
      });
    });

    it('should be disabled', () => {
      render(<Button disabled>Default Disabled</Button>);

      const defaultDisabled = screen.getByRole('button', {
        name: /default disabled/i,
      });

      expect(defaultDisabled).toBeDisabled();
      expect(defaultDisabled).toHaveStyle(DISABLED_STYLE);
    });

    it('should be focused', () => {
      render(<Button>Default Focused</Button>);

      const defaultFocused = screen.getByRole('button', {
        name: /default focused/i,
      });

      defaultFocused.focus();

      expect(defaultFocused).toHaveFocus();
      expect(defaultFocused).toHaveStyleRule('background-color', '#aeaeae', {
        modifier: ':focus',
      });
    });

    it('should be hovered', () => {
      render(<Button>Default Hovered</Button>);

      expect(
        screen.getByRole('button', { name: /default hovered/i })
      ).toHaveStyleRule('background-color', '#aeaeae', {
        modifier: ':hover',
      });
    });

    it('should display an icon', () => {
      render(<Button children={<InsertEmoticon title="Emoticon" />} />);

      expect(screen.getByRole('img')).toBeInTheDocument();
    });

    it('should display an icon on the left', () => {
      render(
        <Button leftIcon={<InsertEmoticon title="Emoticon" />}>
          Left Icon
        </Button>
      );

      const withLeftIcon = screen.getByRole('img');

      expect(withLeftIcon).toBeInTheDocument();
      expect(withLeftIcon).toHaveStyle({ 'margin-right': '0.5rem' });
    });

    it('should display an icon on the right', () => {
      render(
        <Button rightIcon={<InsertEmoticon title="Emoticon" />}>
          Right Icon
        </Button>
      );

      const withRightIcon = screen.getByRole('img');

      expect(withRightIcon).toBeInTheDocument();
      expect(withRightIcon).toHaveStyle({ 'margin-left': '0.5rem' });
    });
  });

  describe('"outline" type', () => {
    it('should be styled correctly', () => {
      render(<Button variant="outline">Outline</Button>);

      expect(screen.getByRole('button', { name: /outline/i })).toHaveStyle({
        'background-color': '#fff',
        'border-color': '#2962ff',
        color: '#2962ff',
      });
    });

    it('should be disabled', () => {
      render(
        <Button disabled variant="outline">
          Disabled
        </Button>
      );

      const button = screen.getByRole('button', { name: /disabled/i });

      expect(button).toBeDisabled();
      expect(button).toHaveStyle(DISABLED_STYLE);
    });

    it('should be focused', () => {
      render(<Button variant="outline">Outline Focused</Button>);

      const outlineFocused = screen.getByRole('button', {
        name: /outline focused/i,
      });

      outlineFocused.focus();

      expect(outlineFocused).toHaveFocus();
      expect(outlineFocused).toHaveStyleRule(
        'background-color',
        'rgba(41,98,255,0.1)',
        { modifier: ':focus' }
      );
    });

    it('should be hovered', () => {
      render(<Button variant="outline">Hovered</Button>);

      expect(screen.getByRole('button', { name: /hovered/i })).toHaveStyleRule(
        'background-color',
        'rgba(41,98,255,0.1)',
        {
          modifier: ':hover',
        }
      );
    });
  });

  describe('"text" type', () => {
    it('should be styled correctly', () => {
      render(<Button variant="text">Text</Button>);

      expect(screen.getByRole('button', { name: /text/i })).toHaveStyle({
        'background-color': 'transparent',
        'border-color': 'transparent',
        color: '#2962FF',
      });
    });

    it('should be disabled', () => {
      render(
        <Button disabled variant="text">
          Disabled
        </Button>
      );

      const button = screen.getByRole('button', { name: /disabled/i });

      expect(button).toBeDisabled();
      expect(button).toHaveStyle({
        'background-color': '#fff',
        'border-color': 'transparent',
        color: '#9e9e9e',
        'pointer-events': 'none',
      });
    });

    it('should be focused', () => {
      render(<Button variant="text">Text Focused</Button>);

      const textFocused = screen.getByRole('button', {
        name: /text focused/i,
      });

      textFocused.focus();

      expect(textFocused).toHaveFocus();
      expect(textFocused).toHaveStyleRule(
        'background-color',
        'rgba(41,98,255,0.1)',
        { modifier: ':focus' }
      );
    });

    it('should be hovered', () => {
      render(<Button variant="text">Hovered</Button>);

      expect(screen.getByRole('button', { name: /hovered/i })).toHaveStyleRule(
        'background-color',
        'rgba(41,98,255,0.1)',
        {
          modifier: ':hover',
        }
      );
    });
  });

  describe('"link" type', () => {
    it('should be rendered as an "default" variant', () => {
      render(
        <Button as="a" href="/link">
          Link
        </Button>
      );

      const link = screen.getByRole('link', { name: /link/i });

      expect(link).toHaveAttribute('href', '/link');
      expect(link).toHaveStyle({
        'background-color': '#e0e0e0',
        'box-shadow': '0 0.2rem 0.3rem rgba(51,51,51,0.2)',
        color: '#3f3f3f',
      });
    });

    it('should be rendered as an "outline" variant', () => {
      render(
        <Button as="a" href="/link" variant="outline">
          Link
        </Button>
      );

      const link = screen.getByRole('link', { name: /link/i });

      expect(link).toBeInTheDocument();
      expect(link).toHaveStyle({
        'background-color': '#fff',
        'border-color': '#2962ff',
        color: '#2962ff',
      });
    });

    it('should be rendered as a "text" variant', () => {
      render(
        <Button as="a" href="/link" variant="text">
          Text
        </Button>
      );

      const link = screen.getByRole('link', { name: /text/i });

      expect(link).toBeInTheDocument();
      expect(link).toHaveStyle({
        'background-color': 'transparent',
        'border-color': 'transparent',
        color: '#2962FF',
      });
    });

    it('should be disabled', () => {
      render(
        <Button as="a" href="/link" disabled>
          Disabled
        </Button>
      );

      const linkDefaultDisabled = screen.getByRole('link', {
        name: /disabled/i,
      });

      expect(linkDefaultDisabled).toHaveAttribute('href', '/link');
      expect(linkDefaultDisabled).toHaveStyle(DISABLED_STYLE);
    });

    it('should be focused', () => {
      render(
        <Button as="a" href="/link">
          Link Default Focused
        </Button>
      );

      const linkDefaultFocused = screen.getByRole('link', {
        name: /link default focused/i,
      });

      linkDefaultFocused.focus();

      expect(linkDefaultFocused).toHaveFocus();
      expect(linkDefaultFocused).toHaveStyleRule(
        'background-color',
        '#aeaeae',
        { modifier: ':focus' }
      );
    });

    it('should be hovered', () => {
      render(
        <Button as="a" href="/link">
          Hovered
        </Button>
      );

      const linkDefaultHovered = screen.getByRole('link', {
        name: /hovered/i,
      });

      expect(linkDefaultHovered).toHaveAttribute('href', '/link');
      expect(linkDefaultHovered).toHaveStyleRule(
        'background-color',
        '#aeaeae',
        { modifier: ':hover' }
      );
    });
  });
});
