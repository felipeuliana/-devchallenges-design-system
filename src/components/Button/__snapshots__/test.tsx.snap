// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render correctly 1`] = `
.c0 {
  border: 0.1rem solid;
  border-radius: 0.6rem;
  cursor: pointer;
  font-family: 'Noto Sans JP',sans-serif;
  font-size: 1.4rem;
  font-style: normal;
  font-weight: 500;
  line-height: 2rem;
  padding: 0.8rem 1.6rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.25s ease-in-out,border-color 0.25s ease-in-out;
  transition: background-color 0.25s ease-in-out,border-color 0.25s ease-in-out;
  background-color: #2962ff;
  border-color: #2962ff;
  color: #fff;
  padding: 1.1rem 2.2rem;
  box-shadow: 0 0.2rem 0.3rem rgba(51,51,51,0.2);
}

.c0 > svg {
  fill: currentColor;
  height: 2rem;
  margin-right: 0.5rem;
}

.c0:focus,
.c0:hover {
  background-color: #0039cb;
  border-color: #0039cb;
}

.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

<button
  class="c0"
  color="primary"
>
  <svg
    class="c1"
    fill="currentColor"
    focusable="false"
    role="img"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      Emoticon
    </title>
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
    <path
      d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm3.5-9c.83 0 1.5-.67 1.5-1.5S16.33 8 15.5 8 14 8.67 14 9.5s.67 1.5 1.5 1.5zm-7 0c.83 0 1.5-.67 1.5-1.5S9.33 8 8.5 8 7 8.67 7 9.5 7.67 11 8.5 11zm3.5 6.5c2.33 0 4.31-1.46 5.11-3.5H6.89c.8 2.04 2.78 3.5 5.11 3.5z"
    />
  </svg>
  Emoticon
</button>
`;

exports[`<Button /> should render correctly 2`] = `
.c0 {
  border: 0.1rem solid;
  border-radius: 0.6rem;
  cursor: pointer;
  font-family: 'Noto Sans JP',sans-serif;
  font-size: 1.4rem;
  font-style: normal;
  font-weight: 500;
  line-height: 2rem;
  padding: 0.8rem 1.6rem;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.25s ease-in-out,border-color 0.25s ease-in-out;
  transition: background-color 0.25s ease-in-out,border-color 0.25s ease-in-out;
  background-color: transparent;
  border-color: transparent;
  color: #2962ff;
  padding: 0.6rem 1.2rem;
}

.c0 > svg {
  fill: currentColor;
  height: 2rem;
}

.c0:focus,
.c0:hover {
  background-color: rgba(41,98,255,0.1);
}

<a
  class="c0"
  color="default"
  href="/link"
>
  Link
</a>
`;
